### DISCUSSION TITLE
Configuration package improvements (validation and data structure)

### DISCUSSION THREAD

**Discussion by System-Glitch** [go-goyave/goyave#212] [Ideas]
#210 [#211](https://github.com/go-goyave/goyave/pull/211#discussion_r1647610058) started a discussion on inconveniences with the configuration package, especially when it comes to more **complex data structures** (slices of structures for example) and **validation** of the entries. I would like to take some time and think about improvements for the config package, or maybe a redesign for a future major version.

Please share here your wishlist for the configuration package, talk about your use-cases, etc. Feel free to share your ideas!

**Comment by BowlOfSoup** [comment]
Thanks for creating this discussion. I have several use-cases.

**1. Support for slices of structures**
For example; I want to support and configure a set of API clients, which use in the code is based on user input via an API request. The configuration I would like to use can look like this:

```json
"some_config_list_of_endpoints": [
    {
        "name": "dummy entry",
        "api_key": "value123",
        "url": "https://example.com",
    },
    {
        "name": "yet another entry",
        "api_key": "uuid-something-something",
        "url": "https://some-url.com",
    }
]
```

I would like to have the `some_config_list_of_endpoints` as key/category and the specific properties, defined and validated. This will make Goyave more flexible and adaptable for varying config requirements.

**2. Configuration in structs**
To use configuration in certain services we could pass the main `server` or `config` struct and `Get()` config from there, but wouldn't it be great if we could pass specific configuration itself to services/functions to have clear **separation of concerns**? Why would I pass so much (potentially) information around if I just need the `some_config_list_of_endpoints` config? We could 'unmarshal' the config into structs.

* Better separation of concern
* Enhanced type safety
* Improved testability (of code that uses the config)

```go
var EndpointConfigs []EndpointConfig // example

type EndpointConfig struct {
	Name      string `json:"name"`
	ApiKey    string `json:"api_key"`
	BaseUrl   string `json:"url"`
}
```

**3. Content validation for configuration**
Currently (upcoming Goyave 5.1) supports 'required' validation. This means that I can assert that the `name` property of the above config is required and should not be `nil` or be omitted entirely. This is good, but I would like to take it a step further by validating the actual value.

_**3.1 Validation through using a RuleSet (per entry)**_
This can be done like:

```go
config.Register("some_config_list_of_endpoints.name", config.Entry{
	Value: "",
	Type:  reflect.String,
	Validation: v.RuleSet{
		Path: ".", 
		Rules: v.List{v.Required(), v.String(), v.Min(1)},
	}
})
```

This would remove boilerplate, because one does not have to define a separate validation layer in their app for config.

_**3.2 Validation via the eventual struct**_
This idea needs suggestion 2, putting the configuration in defined structs. We could then define validation on the struct and go from there. Example:

```go
type CustomAuthService struct {
	Url string `json:"url" validate:"required, type=string, min=1"`
}
```

This way we circumvent "manually" defining the RuleSets and the need for validation boilerplate. Also this way we centralize defining, validation and containing configuration within Goyave.

We could even look into supporting custom validation libraries like `go-playground/validator` (or build better, because there are some quirks in that lib).

**Generic Pros**
* Goyave's configuration system would be more robust with these changes
* We have an enhanced development experience, develop faster and safer
* Improved flexibility and scalability
* Improved centralization of config logic overall

**Possible cons**
* We need to make sure performance is not negatively impacted
* We should keep backward compatibility in mind (breaking changes).

**Comment by System-Glitch** [comment]
By using an interface instead of `*config.Config` in the validation package, we would remove the hard dependency and could probably use the validation system directly in the config package. Using such interface would still be considered a breaking change though, as the config type available to validators would be different and would be missing the `Set()` method (which they weren't supposed to use though).

**Comment by System-Glitch** [comment]
I gave it more thoughts and I decided to go back to the drawing board and redesign the configuration part of the framework from scratch using your suggestions and the various feedback I gathered from other users. Here are my first notes and ideas:

### Kept characteristics
- The configuration must remain flexible and easily expandable. It must allow easy development and integration of plugins / external libraries.
- The configuration must remain validated. Current validation system is a bit awkward and very rudimentary compared to the `validation` package. 
- Optional entries must be supported.
- Values loaded from environment variables must be supported.

### Type-safe design
- Make the config system entirely type-safe using structures and composition. This custom struct could be defined in a new file `config.go` at the root of the project. For example:
```go
type Config struct{
    config.BaseConfig

    MyService myservice.Config `json:"myService"`
    SomeLib   somelib.Config `json:"someLib"`
}
```
  - `typeutil.Undefined` must be supported for optional config entries. This will also make it easier for consumers to know if the value was provided or not.
- The input config is unmarshalled into a `any` variable just like the parse middleware. Expect anything as input and validate it later.
- Current inferior configuration validation system is dropped in favor of the `validation` package. This package flawlessly handles a very exhaustive list of use-cases and types already. Let's just use the superior solution. It will cover the vast majority of use-cases: from slices of structs to custom typing and more complex criteria (such as allowed values, regex, etc). Developers can chose how in-depth they go with validation of their config.
  - A configuration struct can implement a function returning a ruleset.
- After validation, convert the output into the custom `Config` struct. This is the same logic used to handle request bodies in controllers and DTO conversion.
- This approach feels cleaner and more intuitive than `config.Register`. Now the rules are defined by the struct receiver too, not by globals.
- Distribution:
  - Each "component" (service, repository, or Goyave `Component`) must define its configuration next to its implementation and take it as parameter of their constructor. If they need values from the base config, accept them as additional parameters. If possible, avoid taking the entire base config as parameter, only take what your component needs.
  - The main config object should not be "global" and distributed everywhere. It's not needed anymore if the point above is respected. I am still undecided about if I should still store the entire config in the server or only what's relevant for it. Using custom types for this can make it challenging because I would need to rely on generics everywhere for the config type.
  - This approach makes testing even easier as you won't need to create a entire new config object in every test. Your tested code will only depend on what it strictly needs. 

### New features

- Allow other formats such as yaml and toml.
  - The current best parsers for [yaml](https://github.com/goccy/go-yaml) and [toml](https://github.com/pelletier/go-toml) respect the behavior of `encoding/json` so they would work perfectly fine with already existing systems inside the framework.
  - This would give more options for developers to use the config format they prefer. 
- Make sure the new implementation will be compatible with `encoding/json/v2`. If not available yet at the time of release, it should be possible to upgrade without introducing breaking changes.
- Allow env variable interpolation in string entries. For example `${VAR_1} - ${VAR_2}`
- Allow overriding config entries using command-line flags using a path (just like the paths in rulesets). No idea yet as to how to implement this. This would come with a lot of edge cases (slices, types non-convertible from string, etc)
- Support loading multiple config files and merging the result (#273).

### Performance consideration

Loading the config will probably be slightly slower since more will be happening with the validation and conversion. But overall I expect the impact to be minor at startup.

It would be much faster at runtime though, since the config will now be simply accessed by reference. No more need for path parsing and multiple map lookups.

### Caveats / Things to clarify

- If a custom config structure doesn't implement a ruleset function, unmarshalling into the custom Config struct may fail without proper error.
  - Let it this way. Developers will chose if they want actual validation or if the basic type check of `typeutil.Convert` is enough.
  - Use reflection to create a ruleset automatically. We will need to use reflection anyway to build the full ruleset. Consider all fields required in this case? It would probably be better for DX.
- This approach makes it difficult to add configuration entries into existing categories (categories from the base config). (I'm reusing the current config lingo here)
- This can be a significant breaking change in large projects, especially if they do not already use service-localized configuration structures.
- `Object` validator should still work just fine even with yaml/toml parsers.
- Some built-in features might be slightly less convenient to use because config would need to be provided by the developer whereas before it was automatically retrieved from the "global" config object.
- Setting the config values manually from the config struct bypasses advanced validation. But at least it's type-safe.
- It is a bit unclear right now how env variable loading would work with the new system. Maybe a new validator `config.FromEnv()` could be silently appended to all fields in the rulesets? It would check if the value is a string and replace the matching tokens with the value retrieved from env.

I will need to experiment a bit with this idea of a design.

**Comment by System-Glitch** [comment]
I started working on this feature and made a first working version on the `v6/config` branch: https://github.com/go-goyave/goyave/compare/v6/config

Feedback is always appreciated!
